packages/compiler/src/server/compile-service.ts

```diff
@@ -331,46 +331,71 @@ export function createCompileService({
 
     let dir = getDirectoryPath(path);
     const options = { allowFileNotFound: true };
+    const entrypoints = clientConfigsProvider?.config?.entrypoint;
+    let fallbackCandidate: string | undefined;
 
     while (true) {
-      let mainFile = "main.tsp";
-      let pkg: any;
-      const pkgPath = joinPaths(dir, "package.json");
-      const cached = await fileSystemCache.get(pkgPath);
-
-      if (cached?.data) {
-        pkg = cached.data;
-      } else {
-        [pkg] = await loadFile(
-          compilerHost,
-          pkgPath,
-          JSON.parse,
-          logMainFileSearchDiagnostic,
-          options,
-        );
-        await fileSystemCache.setData(pkgPath, pkg ?? {});
+      if (entrypoints && entrypoints.length > 0) {
+        // Check for client provided entrypoints (highest priority)
+        for (const entrypoint of entrypoints) {
+          const candidate = joinPaths(dir, entrypoint);
+          const stat = await doIO(
+            () => compilerHost.stat(candidate),
+            candidate,
+            logMainFileSearchDiagnostic,
+            options,
+          );
+
+          if (stat?.isFile()) {
+            log({
+              level: "debug",
+              message: `main file found using client provided entrypoint: ${candidate}`,
+            });
+            return candidate;
+          }
+        }
       }
 
-      const tspMain = resolveTspMain(pkg);
-      if (typeof tspMain === "string") {
-        log({
-          level: "debug",
-          message: `tspMain resolved from package.json (${pkgPath}) as ${tspMain}`,
-        });
-        mainFile = tspMain;
-      }
+      if (!fallbackCandidate) {
+        let mainFile = "main.tsp";
+        let pkg: any;
+        const pkgPath = joinPaths(dir, "package.json");
```

Comment:
I think the package.json should still be the boundary regardless of what was configured. If we go higher more things will break.

Reply:
make sense.